services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -86,20 +91,39 @@
     private S3Utilities(Builder builder) {
         this.region = Validate.paramNotNull(builder.region, "Region");
         this.endpoint = builder.endpoint;
-        this.profileFile = builder.profileFile;
+        this.profileFile = builder.profileFile != null ? () -> builder.profileFile
+                                                       : new Lazy<>(ProfileFile::defaultProfileFile)::getValue;
         this.profileName = builder.profileName;
 
         if (builder.s3Configuration == null) {
             this.s3Configuration = S3Configuration.builder().dualstackEnabled(builder.dualstackEnabled).build();
         } else {
-            this.s3Configuration = builder.s3Configuration.toBuilder()
-                                                          .applyMutation(b -> {
-                                                              if (b.dualstackEnabled() == null) {
-                                                                  b.dualstackEnabled(builder.dualstackEnabled);
-                                                              }
-                                                          })
-                                                          .build();
+            this.s3Configuration =
+                builder.s3Configuration.toBuilder()
+                                       .applyMutation(b -> {
+                                           if (b.dualstackEnabled() == null) {
+                                               b.dualstackEnabled(builder.dualstackEnabled);
+                                           }
+
+                                           if (b.dualstackEnabled() == null) {
+                                               b.dualstackEnabled(DualstackEnabledProvider.builder()
+                                                                                          .profileFile(profileFile)
+                                                                                          .profileName(profileName)
+                                                                                          .build()
+                                                                                          .isDualstackEnabled()
+                                                                                          .orElse(false));
```

Comment:
Might be readable if it's something like below imo

```java
if (b.dualstackEnabled() == null) {
    boolean dualstackEnabled = Optional.ofNullable(builder.dualstackEnabled)
                                       .orElseGet(() -> DualStckEnabledProvider.builder()...)
   
    b.dualstackEnabled(dualstackEnabled);
}
```

Reply:
I just tried that out, and it ends up being too much nesting to keep track of in my head at once. I tried breaking it out into other methods and ended up just rewriting it, since I assume this comment was because that it was hard to understand.